lib/experimental/full-page-client-side-navigation.php

```diff
@@ -0,0 +1,60 @@
+<?php
+/**
+ * Registers full page client-side navigation option using the Interactivity API and adds the necessary directives.
+ */
+
+/**
+ * Enqueue the interactivity router script.
+ */
+function _gutenberg_enqueue_interactivity_router() {
+	// Set the navigation mode to full page client-side navigation.
+	wp_interactivity_config( 'core/router', array( 'navigationMode' => 'fullPage' ) );
+	wp_enqueue_script_module( '@wordpress/interactivity-router' );
+}
+
+add_action( 'wp_enqueue_scripts', '_gutenberg_enqueue_interactivity_router' );
+
+/**
+ * Set enhancedPagination attribute for query loop when the experiment is enabled.
+ *
+ * @param array $parsed_block The parsed block.
+ *
+ * @return array The same parsed block with the modified attribute.
+ */
+function _gutenberg_add_enhanced_pagination_to_query_block( $parsed_block ) {
+	if ( 'core/query' !== $parsed_block['blockName'] ) {
+		return $parsed_block;
+	}
+
+	$parsed_block['attrs']['enhancedPagination'] = true;
+	return $parsed_block;
+}
+
+add_filter( 'render_block_data', '_gutenberg_add_enhanced_pagination_to_query_block' );
+
+/**
+ * Add directives to all links.
+ *
+ * Note: This should probably be done per site, not by default when this option is enabled.
+ *
+ * @param array $content The block content.
+ *
+ * @return array The same block content with the directives needed.
+ */
+function _gutenberg_add_client_side_navigation_directives( $content ) {
+	$p = new WP_HTML_Tag_Processor( $content );
+	while ( $p->next_tag( array( 'tag_name' => 'a' ) ) ) {
+		if ( empty( $p->get_attribute( 'data-wp-on--click' ) ) ) {
+			$p->set_attribute( 'data-wp-on--click', 'core/router::actions.navigate' );
```

Comment:
For what it is worth, I think we should use event delegation for this. Something similar to this, but adding our checks (if the element already has a `data-wp-on--click`...): https://gist.github.com/devongovett/919dc0f06585bd88af053562fd7c41b7

Reply:
That's a good idea. I guess it makes sense if we address that in a follow-up PR, right?

Reply:
@felixarntz related to the discussion in https://github.com/WordPress/gutenberg/discussions/60574 regarding async event handlers, if event delegation were employed like this above [Gist](https://gist.github.com/devongovett/919dc0f06585bd88af053562fd7c41b7) does, then this is another area where synchronous event handlers can negatively impact INP. It would be likely that analytics plugins would adopt similar event delegation rather than having to use the tag processor to inject their own `click` event directives on each link.

Reply:
For this experiment, I added [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/d72cd023eae40645e8e787f24d91f515b12ce4ee) to use `document.addEventListener` instead of the previous directives. I also added a follow-up task to [this list](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2058832203) to explore this more deeply.